plugins/modules/grafana_folder.py

```diff
@@ -0,0 +1,251 @@
+        if grafana_version["major"] < 5:
+            self._module.fail_json(failed=True, msg="Folders API is available starting Grafana v5")
+
+    def _send_request(self, url, data=None, headers=None, method="GET"):
+        if data is not None:
+            data = json.dumps(data, sort_keys=True)
+        if not headers:
+            headers = []
+
+        full_url = "{grafana_url}{path}".format(grafana_url=self.grafana_url, path=url)
+        resp, info = fetch_url(self._module, full_url, data=data, headers=headers, method=method)
+        status_code = info["status"]
+        if status_code == 404:
+            return None
+        elif status_code == 401:
+            self._module.fail_json(failed=True, msg="Unauthorized to perform action '%s' on '%s' header: %s" % (method, full_url, self.headers))
+        elif status_code == 403:
+            self._module.fail_json(failed=True, msg="Permission Denied")
+        elif status_code == 412:
+            error_msg = resp.read()['message']
+            self._module.fail_json(failed=True, msg=error_msg)
+        elif status_code == 200:
+            return self._module.from_json(resp.read())
+        self._module.fail_json(failed=True, msg="Grafana Folders API answered with HTTP %d" % status_code)
+
+    def get_version(self):
+        url = "/api/health"
+        response = self._send_request(url, data=None, headers=self.headers, method="GET")
+        version = response.get("version")
+        major, minor, rev = version.split(".")
+        return {"major": int(major), "minor": int(minor), "rev": int(rev)}
+
+    def create_folder(self, title):
+        url = "/api/folders"
+        folder = dict(title=title)
+        response = self._send_request(url, data=folder, headers=self.headers, method="POST")
+        return response
+
+    def get_folder(self, title):
+        url = "/api/folders?limit=1000"
+        response = self._send_request(url, headers=self.headers, method="GET")
+        for item in response:
+            if item.get("title") == title:
+                return item
+        return None
+
+    def delete_folder(self, folder_uid):
+        url = "/api/folders/{folder_uid}".format(folder_uid=folder_uid)
+        response = self._send_request(url, headers=self.headers, method="DELETE")
+        return response
+
+
+def setup_module_object():
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=False,
+        required_together=[['url_username', 'url_password']],
+        mutually_exclusive=[['url_username', 'grafana_api_key']],
+    )
+    return module
+
+
+argument_spec = url_argument_spec()
+# remove unnecessary arguments
+del argument_spec['force']
+del argument_spec['force_basic_auth']
+del argument_spec['http_agent']
+
+argument_spec.update(
+    state=dict(choices=['present', 'absent'], default='present'),
+    name=dict(type='str', aliases=['title'], required=True),
+    url=dict(type='str', required=True),
```

Comment:
I suggest to rename this to `grafana_url` like in the other grafana modules (or at least add an alias)

Reply:
Is this option common to all grafana modules? Maybe it should go into the (to be created) module_utils/grafana.py's grafana_arg_spec. I guess that could be done after this PR has been merged. It should simplify future PR development.

Reply:
yes I agree with the need to have a common arg_spec, next step :)